src/NumericalAlgorithms/RootFinding/NewtonRaphson.hpp

```diff
@@ -41,9 +41,11 @@ double newton_raphson(const Function& f, const double initial_guess,
 
   boost::uintmax_t max_iters = max_iterations;
   // clang-tidy: internal boost warning, can't fix it.
-  return boost::math::tools::newton_raphson_iterate(  // NOLINT
+  auto result = boost::math::tools::newton_raphson_iterate(  // NOLINT
       f, initial_guess, lower_bound, upper_bound,
       std::round(std::log2(std::pow(10, digits))), max_iters);
+  ASSERT(max_iters < max_iterations, "Root finder did not converge.");
```

Comment:
You should throw instead since we already allow exceptions from boost to propagate and be caught. This will also allow us to have a backup plan in case something doesn't converge.

Reply:
A custom exception sounds good to me for algorithmic failures that we potentially can work around with a backup plan

Reply:
Okay. I like where this idea is heading, but it's also growing into a larger change that IMO should be a different PR from this bugfix. I'll remove the commit from here and open a new PR when ready.

Reply:
I'd be fine with it being a separate commit or a separate PR, whatever you prefer :)

Reply:
I'll make a separate PR. I've removed the relevant commit from this PR, so it should be ready for you to look at my fixups to your previous review :)